### PR TITLE
pkg: enable dark theme for PF

### PR DESCRIPTION
**Pull request by jelly**
To do after merge:

- [ ] bump COCKPIT_REPO_COMMIT in machines and create a PR
- [ ] bump COCKPIT_REPO_COMMIT in podman and create a PR

---

# Dark mode support

Cockpit now switches to a dark theme when system dark mode is enabled.

### CODE REVIEW

**Comment by garrett** [comment]
Both @jelly and I tried for a while today to get this to work by adding various files in different orders in different places... and it's not working right for either of us.

Perhaps there's some special magic to get it to work as expected that we're just not aware of at the moment?

---

We did also try to add in the following files in addition to those in this PR:

- @patternfly/patternfly/patternfly-theme-dark-prefers-color-scheme.scss
- @patternfly/patternfly/themes/dark/base.css

Nothing's working. Some widgets do show up a bit different (no shadows, darker borders) and sometimes we can get some to change (like white text on a white background when in dark mode), but it's nowhere near a dark mode yet. So we must be doing something wrong.

**Comment by jelly** [comment]
Rebased the branch, and now we get something:

image

**Comment by jelly** [comment]
Mostly FYI:

* When `pf-theme-dark` is set, patternfly enables the dark theme.
* In chrom(ium) toggling dark mode can be done in devtools: click the three dotted (kebab) menu, `more tools` -> `Rendering`. Then select the emulation mode from "Emulate CSS media feature prefers-color-scheme"

**Comment by jelly** [comment]
Last rebase has imporved the situation some obvious issues:

- metrics page has issues

image

image

- top bar looks weird:

image


- Graphs are not readable

image

**Comment by garrett** [comment]
Oh, you didn't push your rebase... so, as it rebased cleanly, I just did. :wink:

**Comment by garrett** [comment]
One huge issue right now: On page load, the page is white. That's a dealbreaker for a proper dark theme. I guess we can set html or body to `#000` on load (possibly even in the page, in a `<style>` block, if we can't get to it fast enough otherwise) and then let PF set it to whatever.

**Comment by jelly** [comment]
> Oh, you didn't push your rebase... so, as it rebased cleanly, I just did. wink

Hmm strange, I force pushed again with a rebase plus fixes for some pages which did not have pf-dark-theme set.

**Comment by jelly** [comment]
Pushed one fix for the graphs in storaged/networkmanager page.

**Comment by jelly** [comment]
Update on the flashing between loading pages:

https://drive.google.com/file/d/1pCrXWyTEAvHmHyY1HGGRpXHFyDmEGHC1/view?usp=sharing

I changed the newly created iframe's background to orange and you can see that's what's causing the flashing.

```
diff --git a/pkg/shell/base_index.js b/pkg/shell/base_index.js
index 46447f0b0..4c7282d1c 100644
--- a/pkg/shell/base_index.js
+++ b/pkg/shell/base_index.js
@@ -145,11 +145,13 @@ function Frames(index, setupIdleResetTimers) {
         /* Need to create a new frame */
         if (!frame) {
             new_frame = true;
+            console.log('new frame creation');
             frame = document.createElement("iframe");
             frame.setAttribute("class", "container-frame");
             frame.setAttribute("name", name);
             frame.setAttribute("data-host", host);
             frame.style.display = "none";
+            frame.style.background = 'orange';

             let base, checksum;
             if (machine) {
```

**Comment by garrett** [comment]
I pushed up a bunch of fixes for dark mode. I'm still ignoring the issues that a newer PF would fix; these are ones that affect us specifically and ones that PF has not addressed.

While I did solve a few things already in Metrics, it still has white lines that I've been trying to combat. I wasn't able to figure it out in time, before heading off for a few days of PTO. I'll see you next week!

**Comment by garrett** [comment]
Rebased

**Comment by garrett** [comment]
Rebased & stacked more fixes on top.

I couldn't find what I had done prior, but realized it was simpler than what I had done before, so I just fixed it here as-is.

But I think it's probably good enough for a beta now?! (We should probably reorder and squash a bit more and then merge this in.)

**Comment by jelly** [comment]
The unit-tests fail, investiating

File: pkg/shell/nav.scss [hunk 1]
```diff
@@ -37,9 +58,9 @@ $desktop: $phone + 1px;
     padding: var(--pf-global--spacer--md) var(--pf-global--spacer--md) var(--pf-global--spacer--sm);
 
     > .pf-c-text-input-group__main {
-      background: var(--ct-color-nav);
-      color: var(--ct-color-nav-text) !important;
-      transition: var(--pf-global--Transition);
+      background: var(--ct-color-nav-search);
+      // color: var(--ct-color-nav-text) !important;
+      // transition: var(--pf-global--Transition);
```
Comment (jelly) [inline pkg/shell/nav.scss:63]: This can be dropped I suppose?

Comment (garrett) [reply pkg/shell/nav.scss:63]: Good catch. Yes.

**Comment by jelly** [comment]
image

Text here seems unreadable.

image

Would this be fixed by updating our machines cockpit lib? (The PF overrides)

**Comment by jelly** [comment]
This is still broken:

image

**Comment by marusak** [comment]
Talked with @garrett about pixel tests for dark theme. We think we could auto test dark theme on all biggest screenshots. We don't necessarily need medium and mobile for dark theme.

@martinpitt @jelly @mvollmer Any concerns/options about this?

**Comment by jelly** [comment]
Pixel tests will fail for the dark theme in the shell nav as the highlight color changed for the light theme:

> the highlight color will still change, but i'll be an orangey yellow

File: pkg/lib/cockpit.js [hunk 2]
```diff
@@ -2699,6 +2714,25 @@ function factory() {
...
+            cockpit.dark_mode = true;
+        } else {
+            cockpit.dark_mode = false;
+        }
+
+        window.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', event => {
+            if (event.matches) {
+                cockpit.dark_mode = true;
+            } else {
+                cockpit.dark_mode = false;
```
Comment (cockpituous) [inline pkg/lib/cockpit.js:2731]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221027-154610-ab00b7d4-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/base_index.js [hunk 3]
```diff
@@ -205,6 +205,12 @@ function Frames(index, setupIdleResetTimers) {
         if (new_frame) {
             list[component] = frame;
             document.getElementById("content").appendChild(frame);
+            // Set background to dark mode
+            if (cockpit.dark_mode && frame.contentDocument && frame.contentDocument.documentElement) {
+                // --pf-global--BackgroundColor--light-200
+                const dark_mode_background = '#0f1214';
+                frame.contentDocument.documentElement.style.background = dark_mode_background;
+            }
```
Comment (cockpituous) [inline pkg/shell/base_index.js:213]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221027-154610-ab00b7d4-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by jelly** [comment]
> Talked with @garrett about pixel tests for dark theme. We think we could auto test dark theme on all biggest screenshots. We don't necessarily need medium and mobile for dark theme.
> 
> @martinpitt @jelly @mvollmer Any concerns/opinions about this?

Yes, the question is how we can test this, I would not want to introduce a new test scenario:

- [ ] does the CDP support setting dark theme?
- [ ] does Chrome allow setting via settings?
- [ ] does firefox allow it via settings?

**Comment by jelly** [comment]
> > Talked with @garrett about pixel tests for dark theme. We think we could auto test dark theme on all biggest screenshots. We don't necessarily need medium and mobile for dark theme.
> > @martinpitt @jelly @mvollmer Any concerns/opinions about this?
> 
> Yes, the question is how we can test this, I would not want to introduce a new test scenario:
> 
> * [x]  does the CDP support setting dark theme?
> * [ ]  does Chrome allow setting via settings?
> * [ ]  does firefox allow it via settings?

We can use:

https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#method-setEmulatedMedia

File: pkg/lib/cockpit.js (hunk 2 above)
Comment (cockpituous) [inline pkg/lib/cockpit.js:2731]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221028-082204-6abb7d32-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/base_index.js (hunk 3 above)
Comment (cockpituous) [inline pkg/shell/base_index.js:213]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221028-082204-6abb7d32-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by jelly** [comment]
> > > Talked with @garrett about pixel tests for dark theme. We think we could auto test dark theme on all biggest screenshots. We don't necessarily need medium and mobile for dark theme.
> > > @martinpitt @jelly @mvollmer Any concerns/opinions about this?
> > 
> > 
> > Yes, the question is how we can test this, I would not want to introduce a new test scenario:
> > 
> > * [x]   does the CDP support setting dark theme?
> > * [ ]   does Chrome allow setting via settings?
> > * [ ]   does firefox allow it via settings?
> 
> We can use:
> 
> https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#method-setEmulatedMedia

Created a draft PR here https://github.com/cockpit-project/cockpit/pull/17854

I don't want to push more work on to this PR.

File: pkg/lib/cockpit.js (hunk 2 above)
Comment (cockpituous) [inline pkg/lib/cockpit.js:2731]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221028-143606-093d853b-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/base_index.js (hunk 3 above)
Comment (cockpituous) [inline pkg/shell/base_index.js:213]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221028-143606-093d853b-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit.js (hunk 2 above)
Comment (cockpituous) [inline pkg/lib/cockpit.js:2731]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221028-161741-efc4d3f1-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/base_index.js (hunk 3 above)
Comment (cockpituous) [inline pkg/shell/base_index.js:213]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221028-161741-efc4d3f1-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/cockpit.js (hunk 2 above)
Comment (cockpituous) [inline pkg/lib/cockpit.js:2731]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221102-131822-0f1acac0-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/base_index.js (hunk 3 above)
Comment (cockpituous) [inline pkg/shell/base_index.js:213]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221102-131822-0f1acac0-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by jelly** [comment]
New dark theme pixels https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221102-131822-0f1acac0-fedora-36/log.html

**Comment by jelly** [comment]
All looks good to me, this one has a small minor issue, maybe not even dark theme related 
image

**Comment by garrett** [comment]
> All looks good to me, this one has a small minor issue, maybe not even dark theme related

I think it's not dark theme related, but it's probably not seen in light mode... I think in light, it's a table with a white background on modal dialog with a white background.

But it's not a big deal. I think everything I've seen in the pixel tests are at the level of a cosmetic glitch and worthy of a follow-up. I'm fine with merging this, calling it "beta", and fixing whatever small bits we uncover (as well as these tiny things) after the release.

**Comment by jelly** [comment]
@garrett I think the release announcement needs a screenshot!

File: pkg/lib/cockpit.js (hunk 2 above)
Comment (cockpituous) [inline pkg/lib/cockpit.js:2731]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221102-165757-95891920-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/base_index.js (hunk 3 above)
Comment (cockpituous) [inline pkg/shell/base_index.js:213]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17319-20221102-165757-95891920-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by garrett** [comment]
> I think the release announcement needs a screenshot!

Good point!

We should probably take the screenshot using Cockpit Client, which has had dark mode for quite some time already. :wink:

**Comment by garrett** [comment]
I've included an MP4 which might not work on our site, unless we do special things.

Here's a webm:

[darkmode2.webm](https://user-images.githubusercontent.com/10246/199786211-55afd5b6-6aab-44a5-bf4c-d135db8c1ebe.webm)

Original screenshots, if all else fails:

light

dark

**Pull request merged**